evaluator/src/main/kotlin/Rule.kt

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.evaluator
+
+import com.here.ort.model.OrtIssue
+import com.here.ort.model.Severity
+
+/**
+ * The base class for an evaluator rule. Rules use a set of matchers to determine if they apply, and create a list of
+ * errors which are added to the [ruleSet].
+ */
+abstract class Rule(
+    /**
+     * The [RuleSet] this rule belongs to.
+     */
+    val ruleSet: RuleSet,
+
+    /**
+     * The name of this rule.
+     */
+    val name: String
+) {
+    private val ruleMatcherManager = RuleMatcherManager()
+    private val matchers = mutableListOf<RuleMatcher>()
+
+    /**
+     * The list of all issues created by this rule.
+     */
+    val issues = mutableListOf<OrtIssue>()
+
+    /**
+     * Return a human-readable description of this rule.
+     */
+    abstract fun describeRule(): String
+
+    /**
+     * Run this rule by checking if all matchers apply. If this is the case all [issues] configured in this rule are
+     * added to the [ruleSet].
+     */
+    open fun run(): Boolean {
+        println(describeRule())
+
+        val matches = matchers.all { matcher ->
+            matcher.matches().also { matches ->
+                println("\t${matcher.describe()} == $matches")
+                if (!matches) println("\tRule skipped.")
+            }
+        }
+
+        if (matches) {
+            ruleSet.issues += issues
+        }
+
+        return matches
```

Comment:
I find it a bit odd that this base rule's `run()` doesn't actually run anything, but does something else. How about splitting this into two functions `matches()` and `run()`? Also, if the convention is that derived rules always have to call `super.run()` like `PackageRule.run()` does, I'd prefer a less error-prone design with some `runInternal()` (or so) function to implement:

    Rule {
        fun run() {
            if (matches()) runInternal()
        }

        protected abstract fun runInternal()
    }

`runInternal()` is never called from outside, but only the `run()` wrapper, which ensures a rule is never run without also super class matchers to be checked. Maybe a similar `matchesInternal()` approach is also needed.

Reply:
FYI, we already use a similar approach at https://github.com/heremaps/oss-review-toolkit/blob/bc8617b3bf054d64803084e71b27091463d0a81d/downloader/src/main/kotlin/VersionControlSystem.kt#L270-L281 to ensure sub-classes can only implement additional checks, but not override ones from the base class.

Reply:
Changed it almost as you suggested, but I kept adding of the issues to the rule set inside the default function, as this is core functionality. If we need to have rules working differently we could extract an interface later on.